File: remote/src/main/java/lt/markmerkk/JiraConnector.java
package lt.markmerkk;

import java.util.List;
import java.util.Map;
import net.rcarz.jiraclient.BasicCredentials;
import net.rcarz.jiraclient.JiraClient;
import net.rcarz.jiraclient.JiraException;
import net.rcarz.jiraclient.WorkLog;
import rx.Observable;
import rx.Subscriber;

import static org.apache.commons.lang.StringUtils.isEmpty;

/**
 * Created by mariusmerkevicius on 1/29/16.
 * Responsible for validating credengials and connecting
 * {@link JiraClient}
 */
public class JiraConnector implements Observable.OnSubscribe<JiraClient> {

  String hostname, username, password;

  public JiraConnector(String hostname, String username, String password) {
    this.hostname = hostname;
    this.username = username;
    this.password = password;
  }

  @Override
  public void call(Subscriber<? super JiraClient> subscriber) {
    try {
      if (isEmpty(hostname)) throw new JiraException("Error in JIRA address! Please fill in hostname for JIRA!");
      if (isEmpty(username)) throw new JiraException("Username is empty! Please fill in username / password");
      if (isEmpty(password)) throw new JiraException("Password is empty! Please fill in username / password");

      BasicCredentials creds = new BasicCredentials(username, password);
      JiraClient jira = new JiraClient(hostname, creds);

      subscriber.onNext(jira);
      subscriber.onCompleted();
    } catch (JiraException e) {
      subscriber.onError(e);
    }
  }
}

File: remote/src/main/java/lt/markmerkk/JiraConnector.kt
package lt.markmerkk

import net.rcarz.jiraclient.BasicCredentials
import net.rcarz.jiraclient.JiraClient
import net.rcarz.jiraclient.JiraException
import rx.Observable
import rx.Subscriber

/**
 * Created by mariusmerkevicius on 1/29/16.
 * Responsible for validating credengials and connecting
 * [JiraClient]
 */
class JiraConnector(
        val hostname: String?,
        val username: String?,
        val password: String?
) : Observable.OnSubscribe<JiraClient> {
    override fun call(subscriber: Subscriber<in JiraClient>) {
        try {
            if (hostname == null) throw IllegalStateException("Hostname cannot be empty")
            if (hostname.length == 0) throw IllegalStateException("Hostname cannot be empty")
            if (username == null) throw IllegalStateException("Username cannot be empty")
            if (username.length == 0) throw IllegalStateException("Username cannot be empty")
            if (password == null) throw IllegalStateException("Password cannot be empty")
            if (password.length == 0) throw IllegalStateException("Password cannot be empty")
            val jira = JiraClient(hostname, BasicCredentials(username, password))

            subscriber.onNext(jira)
            subscriber.onCompleted()
        } catch (e: IllegalStateException) {
            subscriber.onError(e)
        } catch (e: JiraException) {
            subscriber.onError(e)
        }
    }
}
